mentat/auto_context.py

```diff
@@ -0,0 +1,150 @@
+
+"""
+Context selection is similar to the Knapsack Problem: given a set of items (features),
+and a knapsack (context size), choose the best subset of items that fit. In our case we
+can also select which level include. We implement two methods:
+
+- Greedy: add features one-by-one, at the most detailed level possible, until the maximum
+    token limit is reached.
+
+- LLM: create a full-context of code using the greedy method, then ask the LLM to select
+    only the relevant features. Return its output, up to the token limit.
+"""
+
+
+def select_features_greedy(
+    features: list[CodeFeature],
+    max_tokens: int,
+    model: str = "gpt-4",
+    levels: list[CodeMessageLevel] = [],
+) -> list[CodeFeature]:
+    """Use the greedy method to return a subset of features max_tokens long"""
+    output = list[CodeFeature]()
+    remaining_tokens = max_tokens
+    for feature in features:
+        _levels = list(set(levels) | {feature.level})
+        _levels = sorted(list(_levels), key=lambda v: v.rank)
+        for level in _levels:
+            feature.level = level
+            if feature.count_tokens(model) <= remaining_tokens:
+                output.append(feature)
+                remaining_tokens -= feature.count_tokens(model)
+                break
+
+    return output
+
+
+feature_selection_model = "gpt-4"
+feature_selection_max_tokens = 8192
+feature_selection_temperature = 0.5
+feature_selection_response_buffer = 500
+feature_selection_prompt = dedent("""\
+    You are part of an automated coding system. 
+    Below you will see the line 'User Query:', followed by a user query, followed by the line 'Code Files:' and then a pre-selected subset of a codebase.
+    Your job is to select portions of the code which are relevant to answering that query.
+    The process after you will read the lines code you select, and then return a plaintext plan of action and a list of Code Edits.
+    {{training_prompt}}
+    Each item in the Code Files below will include a relative path and line numbers.
+    Identify lines of code which are relevant to the query.
+    Return a json-serializable list of relevant items following the same format you receive: <rel_path>:<start_line>-<end_line>,<start_line>-<end_line>.
+    It's important to include lines which would be edited in order to generate the answer as well as lines which are required to understand the context.
+    It's equally important to exclude irrelevant code, as it has a negative impact on the system performance and cost.
+    For example: if a question requires creating a new method related to a class, and the method uses an attribute of that
+    class, include the location for the edit as well as where the attribute is defined. If a typing system is used, include
+    the type definition as well, and the location of the expected import.
+    Make sure your response is valid json, for example:
+    ["path/to/file1.py:1-10,53-60", "path/to/file2.py:10-20"]
```

Comment:
This is just an idea for something to try in the future: We could experiment with showing one file at a time (or rather per prompt, we could do it simultaneously) and asking if this file has relevant information for achieving the user request. Upsides:
1. The simpler we keep it the more likely gpt-3.5 is to be able to do it.
2. We can ask about more files than fit in context

Downsides:
1. Maybe seeing everything will help understand what is necessary to understand
2. We have to pay for the user request every time

Reply:
Another downside is latency as we'd make multiple calls, but we could do them in parallel